docs/deprecation-process.md

```diff
@@ -0,0 +1,309 @@
+especially if there are special usage patterns that were not anticipated.
+
+Once things are stable, proceed with the removal.
+Depending on the change, it can be implemented:
+
+* in the next minor release,
+* after a few minor releases,
+* in the next major release,
+* after a few major releases
+
+following the deprecation.
+
+The more disruptive the change is,
+the more time users will need to adjust,
+so trivial deprecations can be removed right away,
+while complex ones can take much longer.
+
+In any case, never implement the deprecation and the removal in the same
+release: wait.
+
+### Prepare removal
+
+Implement the removal issue with a PR.
+
+Remove all the deprecated code.
+
+Remove all the tooling (compiling options) related to the deprecated code,
+if any.
+
+Remove all the relevant entries in `DEPRECATED`.
+
+Add a `CHANGELOG` entry for the removal.
+
+Get the removal PR reviewed and merged to the main branch.
+
+### Release removal
+
+When making a new opentelemetry-cpp release,
+in addition to the regular release tasks:
+
+* In the `CHANGELOG` and the release notes, add references
+  to the new removed items for this release.
+
+## Technical guidelines
+
+### CMake options
+
+Assume the option `WITH_FOO` needs to be deprecated.
+
+Code using `WITH_FOO=OFF` or `WITH_FOO=ON` must build as before,
+yet users should be notified if they use `WITH_FOO` in their build.
+
+Assume a build option `WITH_NO_DEPRECATION` already exists.
+
+In a normal build, used in production, code is compiled with
+`WITH_NO_DEPRECATION=OFF`.
+
+In a verification build, code is compiled with `WITH_NO_DEPRECATION=ON`.
+
+Implement the following logic in CMake:
+
+```cmake
+  option(WITH_FOO "DEPRECATED - With the foo feature" OFF)
+
+  if(WITH_FOO)
+    if(WITH_NO_DEPRECATION)
+      message(FATAL_ERROR "WITH_FOO is deprecated")
+    else()
+      message(WARNING "WITH_FOO is deprecated")
+    endif()
+  endif
+```
+
+The verification build is not meant to be used in production.
+It is meant to find all references of deprecated items.
+
+If the verification build succeeds, the user code is guaranteed to be clean,
+and will not be impacted by the removal to come.
+
+When implementing such logic, document it in the mitigation section,
+in file `DEPRECATED`.
+
+### C++ deprecation
+
+Assume a C++ item needs to be deprecated.
+
+For example:
+
+```cpp
+  struct some_options
+  {
+    int x;
+    int y; // TODO: deprecate me
+    int z;
+  };
+```
+
+Code using y must build as before, and yet users should be notified if still
+using y.
+
+Implement the following change:
+
+```cpp
+  struct some_options
+  {
+    int x;
+#ifdef HAVE_DEPRECATED_CODE
+    int y; // deprecated, to be removed
+#endif /* HAVE_DEPRECATED_CODE */
+    int z;
+  };
+```
+
+A regular build, to use in production, will build with
+`HAVE_DEPRECATED_CODE` defined.
+
+A verification build, to _not_ use in production,
+will build with `HAVE_DEPRECATED_CODE` undefined.
+
```

Comment:
So we existing CI github actions will now be built with `HAVE_DEPRECATED_CODE` defined. And we will add a new github action to build without `HAVE_DEPRECATED_CODE` ?

Reply:
> Thanks, a very well thought-out and documented process. This will indeed smoothen the deprecation process. Have few nit comments :)

Thanks @lalitb for the comments. I will reword and add clarifications for the remaining points.

Reply:
We can have CI build without the deprecated code, sure (I used the maintainer_mode for that).

The main point here is that users, building an application that depends on opentelemetry-cpp, should:

- make a regular build as usual, to use in their project
- **make a verification build of their own code, to identify deprecated items still in use in their own code.**

Only users can assess if their application uses a given api or not, we can only provide helper tools for that.